common/flatpak-run-wayland.c

```diff
@@ -30,6 +33,44 @@
 
 #include "flatpak-utils-private.h"
 
+static char *
+get_wayland_socket_path (void)
+{
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  const char *wayland_display;
+  char *wayland_socket = NULL;
+
+  wayland_display = g_getenv ("WAYLAND_DISPLAY");
+  if (!wayland_display)
+    wayland_display = "wayland-0";
+
+  if (wayland_display[0] == '/')
+    wayland_socket = g_strdup (wayland_display);
+  else
+    wayland_socket = g_build_filename (user_runtime_dir, wayland_display, NULL);
+  return wayland_socket;
+}
+
+static int
+get_wayland_socket_fd (void)
+{
+  const char *wayland_socket_fd;
+  char *end;
+  int fd = -1;
+  wayland_socket_fd = g_getenv ("WAYLAND_SOCKET");
+  if (!wayland_socket_fd)
+    return -1;
+
+  int prev_errno = errno;
+  errno = 0;
```

Comment:
What is this about?

Reply:
@wengxt, please don't mark conversations as resolved without justifying why your code was correct.

I believe the intention here is that you're following the pattern described in the `strtol()` man page to detect out-of-range values:

```
       Since  strtol()  can  legitimately  return  0,  LONG_MAX,  or  LONG_MIN
       (LLONG_MAX or LLONG_MIN for strtoll()) on both success and failure, the
       calling  program should set errno to 0 before the call, and then deter‐
       mine if an error occurred by checking whether errno has a nonzero value
       after the call.
```

... and if so, that's valid and correct.

However, because `strtol()` is locale-sensitive, please use `g_ascii_strtoll()` or `g_ascii_strtoull()` instead (they behave very similarly).

Also, assigning the result of `strtol()` to an `int` isn't really right, because on LP64 platforms, that will be silently truncating from 64-bit `long` to 32-bit `int`. If the value of `WAYLAND_SOCKET` is outside the 32-bit range of an `int`, that will give the wrong answer with no indication that anything was wrong.

And, this should really be checking that the fd is in the valid range for a fd, which is 0 to `G_MAXINT` inclusive.

Putting all that together, it should look be something like this:

```
guint64 value;

...

errno = 0;
value = g_ascii_strtoull (wayland_socket_fd, &end, 10);
if (errno != 0 || wayland_socket_fd == end || *end != '\0' || value > G_MAXINT)
  return -1;

return (int) value;
```

(`option_env_fd_cb()` elsewhere in the Flatpak codebase *almost* gets this right, although it's still not quite right either.)

Reply:
In fact, even better than what I said before: please use `g_ascii_string_to_unsigned (wayland_socket_fd, 10, 0, INT_MAX, &value, NULL)` which handles all of this noise for you. We have a backport of it in `common/flatpak-glib-backports-private.h` which will be used if GLib is too old.

Reply:
thanks, didn't aware of g_ascii_string_to_unsigned. I know strtol is locale aware and I originally use strtol simply because libwayland-client uses it.